File: python/test/ShapeTest.py
import sys
import unittest
sys.path.append("../")

from pni.utils import Shape
from pni.utils import Index



#implementing test fixture
class ShapeTest(unittest.TestCase):
    def setUp(self):
        self.nx = 100;
        self.ny = 5;
        self.nz = 27;

    def tearDown(self):
        pass

    def test_constructors(self):
        #test default constructor
        s = Shape()
        self.assertTrue(s.rank == 0)
        self.assertTrue(s.size == 0)
        
        s.rank = 2
        s.dim(0,self.nx)
        s.dim(1,self.ny)
        self.assertTrue(s.dim(0) == self.nx)
        self.assertTrue(s.dim(1) == self.ny)
        self.assertTrue(s.size == self.nx*self.ny)

        #test standard constructor
        s2 = Shape(2)
        self.assertTrue(s2.rank == 2)
        self.assertTrue(s2.dim(0) == 0)
        self.assertTrue(s2.dim(1) == 0)

        s2.dim(0,self.ny)
        s2.dim(1,self.nx)
        self.assertTrue(s2.dim(0) == self.ny)
        self.assertTrue(s2.dim(1) == self.nx)
        self.assertTrue(s2.size == self.ny*self.nx)

        #test copy construction
        s3 = Shape(s2)
        self.assertTrue(s3.rank == s2.rank)
        for i in range(s3.rank):
            self.assertTrue(s3.dim(i) == s2.dim(i))

        self.assertTrue(s3.size == s2.size)

    def test_offset_and_index(self):
        i = Index(2)
        s = Shape(2)
        
        s[0] = 6
        s[1] = 9

        i[0] = 3 
        i[1] = 4 
        self.assertTrue(s.offset(i) == 31)

        i1 = Index(2)
        i1[0] = 1 
        i1[1] = 7 
        s.index(16,i)
        self.assertTrue(i1 == i)
    
    def test_shape_manipulation(self):
        s = Shape(2)

        s[0] = 100; s[1] = 200;
        self.assertTrue(s.size == 100*200)

        s.rank = 3
        self.assertTrue(s.size == 0)
        self.assertTrue(s.rank == 3)
        s[0] = 1; s[1] = 2; s[2] = 4
        self.assertTrue(s.size == 1*2*4)


    def test_container_properties(self):
        s = Shape(3)

        self.assertTrue(len(s) == s.rank)
        s.dim(0,self.nx)
        s.dim(1,self.ny)
        s.dim(2,self.nz)

        self.assertTrue(s[0] == self.nx)
        self.assertTrue(s[1] == self.ny)
        self.assertTrue(s[2] == self.nz)

        s[0] = self.nz
        s[1] = self.ny
        s[2] = self.nx

        self.assertTrue(s[0] == self.nz)
        self.assertTrue(s[1] == self.ny)
        self.assertTrue(s[2] == self.nx)

        def test_index(i):
            return s[i]

        self.assertRaises(UserWarning,test_index,4)

        #initialize with slice
        s = Shape(5)
        s[:] = [1,2,3,4,5]
        self.assertTrue(s[1:4] == [2,3,4])

        

File: test/ShapeTest.cpp
//implementation of the arrayshape test

#include<cppunit/extensions/HelperMacros.h>

#include "Types.hpp"
#include "Shape.hpp"
#include "ShapeTest.hpp"
#include "Index.hpp"

CPPUNIT_TEST_SUITE_REGISTRATION(ShapeTest);

//------------------------------------------------------------------------------
void ShapeTest::setUp(){
    _s1 = {3,4};
    _s2 = {2,3,5};
}

//------------------------------------------------------------------------------
void ShapeTest::tearDown(){
}

//------------------------------------------------------------------------------
void ShapeTest::testConstruction(){
	std::cout<<"void ShapeTest::testConstruction()-----------------------";
	std::cout<<std::endl;
    //default initialization of a shape from a std::vector object
	Shape s(_s1);

	//manual construction
	CPPUNIT_ASSERT(s.rank() == _s1.size());
	for(size_t i=0;i<s.rank();i++){
        CPPUNIT_ASSERT(s[i]==_s1[i]);
	}

	//now we change rank and shape of the object - this should go via 
    //implicit conversion from the std::vector object to the shape
	s = _s2;
	for(size_t i=0;i<s.rank();i++){
		CPPUNIT_ASSERT(_s2[i] == s.dim(i));
		CPPUNIT_ASSERT(s[i] == _s2[i]);
	}

    //using copy construction
	Shape s1(s);

    //check comparison of shape objects
	CPPUNIT_ASSERT(s1 == s);

    //construct from an initializer list
    Shape s3 = {10,30,12};
    CPPUNIT_ASSERT(s3[0] == 10);
    CPPUNIT_ASSERT(s3[1] == 30);
    CPPUNIT_ASSERT(s3[2] == 12);
    CPPUNIT_ASSERT(s3.rank() == 3);

    //using move construction
    Shape s4 = std::move(s);
    CPPUNIT_ASSERT(s4.rank() == _s2.size());
    CPPUNIT_ASSERT(s4.size() == s1.size());
    CPPUNIT_ASSERT(s.rank() == 0);
    CPPUNIT_ASSERT(s.size() == 0);

    //reset dimensions using an initializer list
    s4.dim({1,2,3});
    CPPUNIT_ASSERT(s4.size()==6);

}

//------------------------------------------------------------------------------
void ShapeTest::testAssignment(){
	std::cout<<"void ShapeTest::testAssignment()------------------------------";
	std::cout<<std::endl;
	//testing the assignment operator
	Shape s1 = {1,2,3,4};
    Shape s2,s3;

    //copy assignment
    CPPUNIT_ASSERT_NO_THROW(s2 = s1);
    CPPUNIT_ASSERT(s2 == s1);
    CPPUNIT_ASSERT_NO_THROW(s3 = std::move(s1));
    CPPUNIT_ASSERT(s3 == s2);
    CPPUNIT_ASSERT(s3 != s1);
    CPPUNIT_ASSERT(s1.rank() == 0);
    CPPUNIT_ASSERT(s1.size() == 0);

	Shape snon;
	CPPUNIT_ASSERT(s1 == snon);

    //this is done implicitely via the conversion constructor
    snon = {1,2,3,4};
    CPPUNIT_ASSERT(snon.rank() == 4);
    CPPUNIT_ASSERT(snon[0] == 1);
    CPPUNIT_ASSERT(snon[1] == 2);
    CPPUNIT_ASSERT(snon[2] == 3);
    CPPUNIT_ASSERT(snon[3] == 4);
}

//------------------------------------------------------------------------------
void ShapeTest::testComparison(){
	std::cout<<"void ShapeTest::testComparison()------------------------------";
	std::cout<<std::endl;
    Shape s1(_s1);
    Shape s2(_s2);

	Shape s3(s2);

	CPPUNIT_ASSERT(s1 == s1);
	CPPUNIT_ASSERT(s2 == s3);
	CPPUNIT_ASSERT(s1 != s2);

	//what happens with unallocated shapes
	Shape s4,s5;
	CPPUNIT_ASSERT(s4 == s5);
}

//------------------------------------------------------------------------------
void ShapeTest::testOffset(){
	std::cout<<"void ShapeTest::testOffset()----------------------------------";
	std::cout<<std::endl;
	Shape s1(_s1);
	Shape s2(_s2);
	Index index;

	index.rank(s1.rank());
	index[0] = 2;
	index[1] = 1;
	CPPUNIT_ASSERT(s1.offset(index)==9);
    CPPUNIT_ASSERT(s1.offset({2,1}) == 9);

	index.rank(s2.rank());
	index[0] = 0;
	index[1] = 1;
	index[2] = 3;
	CPPUNIT_ASSERT(s2.offset(index)==8);
    CPPUNIT_ASSERT(s2.offset({0,1,3}) == 8);

}

//------------------------------------------------------------------------------
void ShapeTest::testIndex(){
	std::cout<<"void ShapeTest::testIndex()-----------------------------------";
	std::cout<<std::endl;
	Shape s1(_s1),s2(_s2);
	Index index;

	index.rank(s1.rank());
	index[0] = 2;
	index[1] = 1;
	CPPUNIT_ASSERT(s1.offset(index)==9);
	Index ti(s1.rank());
	CPPUNIT_ASSERT_NO_THROW(s1.index(9,ti));
	CPPUNIT_ASSERT(ti == index);

	index.rank(s2.rank());
	ti.rank(s2.rank());
	index[0] = 0;
	index[1] = 1;
	index[2] = 3;
	CPPUNIT_ASSERT(s2.offset(index)==8);
	CPPUNIT_ASSERT_NO_THROW(s2.index(8,ti));
	CPPUNIT_ASSERT(ti == index);

    //testing setting dimensions using a vector
    std::vector<size_t> indices = {5,2,9};
    s2.dim(indices);
    CPPUNIT_ASSERT(s2.size() == 90);
    CPPUNIT_ASSERT(s2.rank() == 3);
    CPPUNIT_ASSERT(s2[0] == 5);
    CPPUNIT_ASSERT(s2[1] == 2);
    CPPUNIT_ASSERT(s2[2] == 9);

}

//------------------------------------------------------------------------------
void ShapeTest::testExceptions(){
	std::cout<<"void ShapeTest::testExceptions()------------------------------";
	std::cout<<std::endl;
	Shape s(_s1);

	CPPUNIT_ASSERT_THROW(s.dim(10,0),IndexError);
	CPPUNIT_ASSERT_THROW(s.dim(-10,0),IndexError);
	CPPUNIT_ASSERT_THROW(s.dim(10),IndexError);
	CPPUNIT_ASSERT_THROW(s.dim(-10),IndexError);
	CPPUNIT_ASSERT_THROW(s[10],IndexError);
	CPPUNIT_ASSERT_THROW(s[-10],IndexError);

    s = _s2;

	Index i(s.rank());
	i[0] = 1; i[1] = 100; i[2] = 1;
	CPPUNIT_ASSERT_THROW(s.offset(i),IndexError);
	i[0] = -1; i[1] = 0;
	CPPUNIT_ASSERT_THROW(s.offset(i),IndexError);

	Index i2(2);
	CPPUNIT_ASSERT_THROW(s.offset(i2),ShapeMissmatchError);
	Index i3;
	CPPUNIT_ASSERT_THROW(s.offset(i3),MemoryAccessError);

	Shape s2;
	CPPUNIT_ASSERT_THROW(s2.offset(i),MemoryAccessError);

	//now we can do the same game with the getIndex method
	//the offset exceeds the size of the shape
	CPPUNIT_ASSERT_THROW(s.index(10000,i),MemoryAccessError);
	//the index has the wrong rank
	CPPUNIT_ASSERT_THROW(s.index(1,i2),ShapeMissmatchError);
	//the shape object is not allocated
	CPPUNIT_ASSERT_THROW(s2.index(1,i),MemoryAccessError);


}
